File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        // failed so many times in this question gonna try again 
        
        int m = matrix.size();
        int n = matrix[0].size();
        
      //  cout<<m<<n;
        int start = 0;
        int end = m*n -1;
        
        while(start<=end){
            int mid = start + (end-start)/2;
            cout<<mid;
            if(matrix[mid/n][mid%n]==target)
                return true;
            
            else if(matrix[mid/n][mid%n]<target){
                start = mid + 1 ;
                cout<<start;
            }
            else {
                end = mid -1 ;
            }
        }
        
        
        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        # return the solution 


        # 0 to 12-1 = 11 

        """
            mid = 5,

            row = 3
            col = 4

            5/4 , 5%4
        """


        low =0
        rows = len(matrix) # 1 
        cols = len (matrix[0]) # 2 
        high = rows*cols-1 # 1 

        while(low<=high):
            mid = (low+high)//2

            # print 

            print(f"mid : {mid}, rows= {rows}")

            if(matrix[mid//cols][mid%cols]==target):
                return True
            
            elif (matrix[mid//cols][mid%cols]<target):
                low = mid+1
            else :
                high= mid -1
        

        return False
